### PR TITLE
Add ability to customize a server display name in the picker list

### PR DESCRIPTION
**Pull request by IanMatthewHuff**
-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for feature-requests.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/notebooks/serviceRegistry.node.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ import { Identifiers } from '../platform/common/constants';
 import { JupyterDebugService } from './debugger/jupyterDebugService.node';
 import { NotebookIPyWidgetCoordinator } from './controllers/notebookIPyWidgetCoordinator';
 import { RemoteKernelConnectionHandler } from './controllers/remoteKernelConnectionHandler';
-import { JupyterServerSelectorCommand } from './serverSelector';
```
Comment (IanMatthewHuff) [inline src/notebooks/serviceRegistry.node.ts:31]: I renamed this as I didn't like how we had two files called 'serverSelector.ts' this file has the command, so name it with just the command.

File: src/kernels/jupyter/types.ts [hunk 2]
```diff
@@ -234,13 +234,21 @@ export interface IJupyterUriProviderRegistration {
...
 export const IJupyterServerUriStorage = Symbol('IJupyterServerUriStorage');
 export interface IJupyterServerUriStorage {
     readonly currentServerId: string | undefined;
     readonly onDidChangeUri: Event<void>;
-    readonly onDidRemoveUris: Event<string[]>;
+    readonly onDidRemoveUris: Event<IJupyterServerUriEntry[]>;
+    readonly onDidAddUri: Event<IJupyterServerUriEntry>;
     addToUriList(uri: string, time: number, displayName: string): Promise<void>;
-    getSavedUriList(): Promise<{ uri: string; serverId: string; time: number; displayName?: string }[]>;
+    getSavedUriList(): Promise<IJupyterServerUriEntry[]>;
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/types.ts:251]: I changed the public interfaces on the URI list to return this whole type (now given a specific type name) from get, onDidRemote, onDidAdd. Felt like it was a good readability improvement and makes it so we don't need to recalculate serverID in a place or two.

File: src/kernels/jupyter/serverSelector.ts [hunk 3]
```diff
@@ -663,3 +676,288 @@ class JupyterServerSelector_Original implements IJupyterServerSelector {
         return items;
     }
 }
+
+/**
+ * Inisders version of the JupyterServerSelector.
+ */
+class JupyterServerSelector_Insiders implements IJupyterServerSelector {
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/serverSelector.ts:683]: This is the new insiders version of the server selector, it's basically the same as the Original version (not 'Experimental'). Don't love the juggling here, but to prototype + work with stable and the previous experiment this seems the best way. I'll call out the different areas in this PR, but it's 98% the same.

File: src/kernels/jupyter/serverSelector.ts [hunk 4]
```diff
@@ -663,3 +676,288 @@ class JupyterServerSelector_Original implements IJupyterServerSelector {
...
+        const uri = await input.showInputBox({
+            title: DataScience.jupyterSelectURIPrompt(),
+            value: initialValue || defaultUri,
+            validate: this.validateSelectJupyterURI,
+            prompt: ''
+        });
+
+        // Offer the user a change to pick a display name for the server
+        // Prepopulate the value with the uri as the default suggestiong
+        const newDisplayName = await this.applicationShell.showInputBox({
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/serverSelector.ts:852]: This is the only big change from the previous server picker, new input box to select a display name. Pre-populated with the URL, so you can just hit enter again to get the previous UI.

File: src/kernels/jupyter/types.ts [hunk 5]
```diff
@@ -234,13 +234,21 @@ export interface IJupyterUriProviderRegistration {
     getJupyterServerUri(id: string, handle: JupyterServerUriHandle): Promise<IJupyterServerUri>;
 }
 
+export interface IJupyterServerUriEntry {
+    uri: string;
+    serverId: string;
+    time: number;
```
Comment (DonJayamanne) [inline src/kernels/jupyter/types.ts:240]: Some comment here would be good, is this time this was added or last accessed?

**Review by DonJayamanne — APPROVED** [review]

File: src/kernels/jupyter/types.ts (hunk 5 above)
Comment (IanMatthewHuff) [reply src/kernels/jupyter/types.ts:240]: Good call. I'll add that in and update.

File: src/kernels/jupyter/types.ts [hunk 6]
```diff
@@ -234,11 +234,12 @@ export interface IJupyterUriProviderRegistration {
...
 
+// An entry into our list of saved servers
 export interface IJupyterServerUriEntry {
-    uri: string;
-    serverId: string;
-    time: number;
-    displayName?: string;
+    uri: string; // URI of the server to connect to
+    serverId: string; // Unique ID using a hash of the full uri
+    time: number; // The most recent time that we connected to this server
```
Comment (DonJayamanne) [inline src/kernels/jupyter/types.ts:241]: Can we add jsDoc style comments
```suggestion
    // The most recent time that we connected to this server
    time: number; 
```

```suggestion
    /** 
     * The most recent time that we connected to this server
     */
    time: number; 
```

These comments show up better in typescript. Else one would have to use `go to definition` to see the above comments.

Comment (IanMatthewHuff) [reply src/kernels/jupyter/types.ts:241]: Bleh, I agree in this case for the definition look up, so I'll change 'em. But I never like how much space the jsdoc style takes up vertically.

Comment (IanMatthewHuff) [reply src/kernels/jupyter/types.ts:241]: Update incoming

**Pull request merged**